plugins/uf2/fu-uf2-device.c

```diff
@@ -0,0 +1,390 @@
+			       GError **error)
+{
+	FuUf2Device *self = FU_UF2_DEVICE(device);
+	gssize wrote;
+	g_autofree gchar *fn = NULL;
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GFile) file = NULL;
+	g_autoptr(GOutputStream) ostr = NULL;
+
+	/* get blob */
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	/* open file for writing; no cleverness */
+	fn = fu_block_device_get_full_path(self, "FIRMWARE.UF2", error);
+	if (fn == NULL)
+		return FALSE;
+	file = g_file_new_for_path(fn);
+	ostr = G_OUTPUT_STREAM(g_file_replace(file, NULL, FALSE, G_FILE_CREATE_NONE, NULL, error));
+	if (ostr == NULL)
+		return FALSE;
+
+	/* write in one chunk and let the kernel do the right thing :) */
+	wrote = g_output_stream_write(ostr,
+				      g_bytes_get_data(fw, NULL),
+				      g_bytes_get_size(fw),
+				      NULL,
+				      error);
+	if (wrote < 0)
+		return FALSE;
+	if ((gsize)wrote != g_bytes_get_size(fw)) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INVALID_FILE,
+			    "only wrote 0x%x bytes",
+			    (guint)wrote);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static GBytes *
+fu_block_device_dump_firmware(FuDevice *device, FuProgress *progress, GError **error)
+{
+	FuUf2Device *self = FU_UF2_DEVICE(device);
+	g_autofree gchar *fn = NULL;
+	g_autoptr(GFile) file = NULL;
+	g_autoptr(GInputStream) istr = NULL;
+
+	/* open for reading */
+	fn = fu_block_device_get_full_path(self, "CURRENT.UF2", error);
+	if (fn == NULL)
+		return NULL;
+	file = g_file_new_for_path(fn);
+	istr = G_INPUT_STREAM(g_file_read(file, NULL, error));
+	if (istr == NULL)
+		return NULL;
+
+	/* read all in one big chunk */
+	return fu_common_get_contents_stream(istr, G_MAXUINT32, error);
+}
+
+static FuFirmware *
+fu_uf2_device_read_firmware(FuDevice *device, FuProgress *progress, GError **error)
+{
+	g_autoptr(FuFirmware) firmware = fu_uf2_firmware_new();
+	g_autoptr(GBytes) fw = NULL;
+
+	fw = fu_device_dump_firmware(device, progress, error);
+	if (fw == NULL)
+		return NULL;
+	if (!fu_firmware_parse(firmware, fw, FWUPD_INSTALL_FLAG_NONE, error))
+		return NULL;
+
+	return g_steal_pointer(&firmware);
+}
+
+static gboolean
+fu_uf2_device_is_mounted_cb(FuDevice *self, gpointer user_data, GError **error)
+{
+	const gchar *devfile = fu_udev_device_get_device_file(FU_UDEV_DEVICE(user_data));
+	g_autoptr(FuVolume) volume = NULL;
+
+	/* mount volume if required */
+	volume = fu_common_get_volume_by_device(devfile, error);
+	if (volume == NULL)
+		return FALSE;
+	if (!fu_volume_is_mounted(volume)) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "is not mounted");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_uf2_device_open(FuDevice *device, GError **error)
+{
+	/* wait for the user session to auto-mount the volume --  we do not want to use
+	 * fu_volume_mount() here and make the volume only accessible by the fwupd user */
```

Comment:
if it's not mounted, why not?  Just take note that you mounted it, let fwupd do it's thing and unmount it.

Reply:
Okay, I've added something like that; 100% untested but feedback welcome.

Reply:
Seems to work.